pkg/tbtc/dkg.go

```diff
@@ -480,47 +484,180 @@ func (de *dkgExecutor) registerSigner(
 	return signer, nil
 }
 
-// submitDkgResult submits the DKG result to the chain.
-func (de *dkgExecutor) submitDkgResult(
+// publishDkgResult performs the DKG result publication process.
+func (de *dkgExecutor) publishDkgResult(
+	ctx context.Context,
 	dkgLogger log.StandardLogger,
 	seed *big.Int,
 	memberIndex group.MemberIndex,
 	broadcastChannel net.BroadcastChannel,
 	membershipValidator *group.MembershipValidator,
-	result *dkg.Result,
+	dkgResult *dkg.Result,
+	groupSelectionResult *GroupSelectionResult,
+	startBlock uint64,
 ) error {
-	// Set up the publication stop signal that should allow to
-	// perform all the result-signing-related actions and
-	// handle the worst case when the result is submitted by the
-	// last group member.
-	chainConfig := de.chain.GetConfig()
-	publicationTimeout := time.Duration(chainConfig.GroupSize) *
-		dkgResultSubmissionDelayStep
-	publicationCtx, cancelPublicationCtx := context.WithTimeout(
-		context.Background(),
-		publicationTimeout,
-	)
-	// TODO: Call cancelPublicationCtx() when the result is
-	//       available and published and remove this goroutine.
-	//       This goroutine is duplicating context.WithTimeout work
-	//       right now but is here to emphasize the need of manual
-	//       context cancellation.
-	go func() {
-		defer cancelPublicationCtx()
-		time.Sleep(publicationTimeout)
-	}()
-
 	return dkg.Publish(
-		publicationCtx,
+		ctx,
 		dkgLogger,
 		seed.Text(16),
 		memberIndex,
 		broadcastChannel,
 		membershipValidator,
-		newDkgResultSigner(de.chain),
-		newDkgResultSubmitter(dkgLogger, de.chain),
-		result,
+		newDkgResultSigner(de.chain, startBlock),
+		newDkgResultSubmitter(dkgLogger, de.chain, groupSelectionResult, de.waitForBlockFn),
+		dkgResult,
+	)
+}
+
+// executeDkgValidation performs the submitted DKG result validation process.
+// If the result is not valid, this function submits an on-chain result
+// challenge. If the result is valid and the given node was involved in the DKG,
+// this function schedules an on-chain approve that is submitted once the
+// challenge period elapses.
+func (de *dkgExecutor) executeDkgValidation(
```

Comment:
I'd consider extracting this function to `dkg_approve.go` or `dkg_validate.go`, for example.

Reply:
It's part of the `dkgExecutor` that lives in `dkg.go` entirely. Not sure if scattering that component across multiple files will give us any benefits.

Reply:
Ah, ok - I see now `dkg_submit.go` hosts `dkgResultSigner` and `dkg_loop.go` hosts `dkgRetryLoop`. That makes sense!